src/modules/uORB/SubscriptionBlocking.hpp

```diff
@@ -0,0 +1,151 @@
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include "SubscriptionCallback.hpp"
+
+#include <containers/LockGuard.hpp>
+#include <px4_time.h>
+
+namespace uORB
+{
+
+// Subscription with callback that schedules a WorkItem
+template<typename T>
+class SubscriptionBlocking : public SubscriptionCallback
+{
+public:
+	/**
+	 * Constructor
+	 *
+	 * @param meta The uORB metadata (usually from the ORB_ID() macro) for the topic.
+	 * @param interval_us The requested maximum update interval in microseconds.
+	 * @param instance The instance for multi sub.
+	 */
+	SubscriptionBlocking(const orb_metadata *meta, uint32_t interval_us = 0, uint8_t instance = 0) :
+		SubscriptionCallback(meta, interval_us, instance)
+	{
+		pthread_mutexattr_t attr;
+
+		int ret_attr_init = pthread_mutexattr_init(&attr);
+
+		if (ret_attr_init != 0) {
+			PX4_ERR("pthread_mutexattr_init failed, status=%d", ret_attr_init);
+		}
+
+		int ret_mutexattr_settype = pthread_mutexattr_settype(&attr, PTHREAD_PRIO_NONE);
+
+		if (ret_mutexattr_settype != 0) {
+			PX4_ERR("pthread_mutexattr_settype failed, status=%d", ret_mutexattr_settype);
+		}
+
+		int ret_mutex_init = pthread_mutex_init(&_mutex, &attr);
+		{
+			PX4_ERR("pthread_mutex_init failed, status=%d", ret_mutex_init);
+		}
+	}
+
+	virtual ~SubscriptionBlocking()
+	{
+		pthread_mutex_destroy(&_mutex);
+		pthread_cond_destroy(&_cv);
+	}
+
+	void call() override
+	{
+		// signal immediately if no interval, otherwise only if interval has elapsed
+		if ((_interval_us == 0) || (hrt_elapsed_time(&_last_update) >= _interval_us)) {
+			pthread_cond_signal(&_cv);
+		}
+	}
+
+	/**
+	 * Copy the struct if updated.
+	 * @param data The data reference where the struct will be copied.
+	 * @param timeout_us The requested timeout in microseconds, or 0 to wait indefinitely.
+	 *
+	 * @return true only if topic was updated and copied successfully.
+	 */
+	bool updateBlocking(T &data, uint32_t timeout_us = 0)
+	{
+		if (!_registered) {
+			registerCallback();
+		}
+
+		if (updated()) {
+			// copy immediately if updated
+			return copy(&data);
+
+		} else {
+			// otherwise wait
+
+			LockGuard lg{_mutex};
+
+			if (timeout_us == 0) {
+				if (pthread_cond_wait(&_cv, &_mutex) == 0) {
+					return update(&data);
+				}
+
+			} else {
+				// otherwise wait with timeout based on interval
+				struct timespec ts;
+
+				if (px4_clock_gettime(CLOCK_REALTIME, &ts) != 0) {
```

Comment:
Shouldn't this be based on the monotonic clock?

And then we need to set `pthread_condattr_setclock (&cond_attr, CLOCK_MONOTONIC)` but that's probably only possible on Posix but not NuttX.
https://github.com/PX4/Firmware/blob/07d656e971a72d1202651dfd3b4642736fb078d7/platforms/posix/src/px4/common/px4_sem.cpp#L65

Reply:
CLOCK_REALTIME is available on NuttX, but needs to be added to several boards.Trying it now.

Reply:
CLOCK_MONOTONIC hangs pthread_cond_timedwait, which appears to use CLOCK_REALTIME internally.

Reply:
Right but can we use it for Posix then?